### PR TITLE
Allow distribution-based counters in http4s module

### PR DESCRIPTION
**Pull request by gstro**
We noticed that the counters in this http4s module were getting overwritten by different invocations of our lambdas. The timers are configurable to be distribution-based, we should allow counters to be distribution-based as well. Happy to make any changes you recommend! Thanks.

### CODE REVIEW

**Comment by tomasherman** [comment]
Thank you for you contribution. Unfortunately i no longer work in Avast so it might take some time before i get this merged as i no longer have admin rights over this project. I am trying to contact my friends in Avast to help me regain the access

**Comment by tomasherman** [comment]
I am not sure i understand the problem and solution though. It doesn't make much sense to me. The counter concept datadog4s is simply a mapping to counter metric in datadog. If you need to use distribution, [it is already present](https://github.com/avast/datadog4s/blob/981d78216bb051bf312e49e40b35dd03165cc413/code/api/src/main/scala/com/avast/datadog4s/api/metric/Distribution.scala)

You can see metric types here: https://docs.datadoghq.com/metrics/types/?tab=distribution#metric-types

I am not sure if overwriting counter in lambdas has any negative implications or whether the creation of a new timer every time is idempotent. However the correct solution in this case is to fix the creation of counters.

The only reason why we do this for timers is that because timer is an abstraction on top of datadog metric types that provides additional logic.

Let me know if i misunderstood or have further comments :)

**Comment by gstro** [comment]
Hi @tomasherman, thanks for the response! Here's hoping you're able to recover access!

Sorry I wasn't clearer in my description here. I'm aware of the `Distribution` metric in datadog4s (and in fact we've been using them with great results!). What this PR does specifically is to allow the internal counters for `requests_count` and `abnormal_count`, that are captured automatically by the `DefaultMetricsOps` implementation, to be configured as distributions or as regular counters. 

Why this is important for Lambdas (from the [Datadog docs](https://docs.datadoghq.com/serverless/custom_metrics/#understanding-distribution-metrics)):
> When Datadog receives multiple count or gauge metric points that share the same timestamp and set of tags, only the most recent one counts. This works for host-based applications because metric points get aggregated by the Datadog agent and tagged with a unique host tag.

> A Lambda function may launch many concurrent execution environments when traffic increases. The function may submit count or gauge metric points that overwrite each other and cause undercounted results. To avoid this problem, custom metrics generated by Lambda functions are submitted as [distributions](https://docs.datadoghq.com/metrics/distributions/) because distribution metric points are aggregated on the Datadog backend, and every metric point counts.

So in the same way that we want to configure the Timers in the `MetricsOps` as distributions for collection server-side, we want to configure the Counters here as distributions for correct reporting by Datadog as well. Hope this helps!

**Comment by tomasherman** [comment]
ah, makes sense :) sorry i didn't read it too carefully i guess ... i think where you are coming from - let me think about it and get back to you. Luckily i got the access i need now aswell :)

**Comment by tomasherman** [comment]
Hey, just a quick update - i hope to resolve this over the weekend :)

File: code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala [hunk 1]
```diff
@@ -15,10 +16,29 @@ private[http4s] class DefaultMetricsOps[F[_]](
...
+      private val counter = metricFactory.count(aspect)
+
+      override def inc(tags: Tag*): F[Unit] = counter.inc(tags*)
+    }
+    final class DistributionCounter(aspect: String) extends GenericCounter {
+      private val counter = metricFactory.distribution.long(aspect)
+
+      override def inc(tags: Tag*): F[Unit] = counter.record(1L, tags*)
+    }
+  }
```
Comment (tomasherman) [inline code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala:40]: alright, so this is the only thing i would change - please move this to it's own file in order to make it more testable and keep this class shorter and therefore understandable. Other than that i think we can merge this.

**Review by tomasherman — CHANGES_REQUESTED** [review]
In addition to these changes, could you also backport this to a `series/cats-effect2` branch?

File: code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala (hunk 1 above)
Comment (gstro) [reply code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala]: This is done!

**Comment by gstro** [comment]
> In addition to these changes, could you also backport this to a `series/cats-effect2` branch?

Done! #888

File: code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala (hunk 1 above)
Comment (gstro) [reply code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala]: @tomasherman How does this look to you?

Comment (tomasherman) [reply code/http4s/src/main/scala/com/avast/datadog4s/extension/http4s/impl/DefaultMetricsOps.scala]: lgtm, lets do it :) Sorry its been busy few weeks in a new job :)

**Review by tomasherman — APPROVED** [review]

**Comment by tomasherman** [comment]
I am not able to test this easily - so i propose this - lets release this as a snapshot version so that you can give it a try and see that it all works well - WDYT? And after you report back i will release it as a new version. Would that wiork for you?

**Comment by tomasherman** [comment]
the snapshot version is: `30.31.3+169-07d35cc8-SNAPSHOT`

**Comment by gstro** [comment]
Works for me! I'll give the snapshot a try likely on Monday so I'll report back here on Tuesday or Wednesday. 👍

**Comment by gstro** [comment]
Looks good on my end, thanks!

**Comment by tomasherman** [comment]
great, i will try to release it over the weekend properly :) Thanks again for contributing and. testing

**Comment by tomasherman** [comment]
so the release was successful, find the new code in version `0.32.0`

**Comment by gstro** [comment]
Awesome, thanks!

**Pull request merged**